### PR TITLE
Unit testing for firestore functions and Tenant Form 

### PR DESCRIPTION
**Pull request by baudoin-csp**
# Unit Testing for Firestore Functions and Tenant Form Screen

## Introduction

Hello Team! 

This PR introduces unit testing for half of the Firestore functions and the `TenantFormScreen` component. The goal is to ensure the reliability of backend operations and the user experience within the tenant form.

---

# Firestore Functions

### Users
- `createUser`
- `getUser`
- `deleteUser`

### Landlords
- `createLandlord`
- `getLandlord`
- `updateLandlord`
- `deleteLandlord`

### Tenants
- `createTenant`
- `getTenant`
- `updateTenant`
- `deleteTenant`

### Residences
- `createResidence`
- `getResidence`
- `updateResidence`
- `deleteResidence`

## Testing Approach

For each function, unit tests were implemented to verify both normal operations and error handling. These tests ensure that:
- Correct Firestore methods are called with appropriate arguments.
- Functions return expected results under normal conditions.
- Errors are properly thrown and handled for invalid inputs or Firestore operation failures.

## Mocking

To facilitate testing without interacting with the actual Firestore database, a `__mocks__` folder was created. It contains mock implementations for Firebase functions, including:
- `getFirestore`
- `collection`
- `doc`
- `addDoc`
- `getDoc`
- `updateDoc`
- `deleteDoc`
- `getDocs`
- `query`
- `where`
- `arrayUnion`

These mocks ensure that tests are isolated and maintainable, providing a solid foundation for testing Firestore functions. The tests cover a wide range of scenarios, ensuring the reliability of database operations.

---

# Tenant Form Screen

### UI Functionality
- **Rendering**: Verifies that the screen renders correctly.
- **Form Interaction**: Ensures input fields can be filled and validations work as expected.
- **Button States**: Confirms that the submit button is disabled when required fields are empty and enabled when all required fields are filled.

### Firestore Integration
- **Successful Submission**: Verifies that `add_new_tenant` is called with correct parameters and navigation occurs after successful submission.
- **Error Handling**: Ensures appropriate error messages are displayed when `add_new_tenant` fails or throws unexpected exceptions.

### Special Cases
- Handles scenarios where a generic error is thrown when a non-Error object is returned.
- Verifies that the "Upload university proof" button does not trigger any unexpected side effects.

## Mocking

In addition to the Firestore function mocks:
- Navigation methods from `@react-navigation/native` were mocked to simulate navigation without actual routing.
- `Alert.alert` was mocked to test user notifications.

These tests provide comprehensive coverage for the `TenantFormScreen`, ensuring both user experience and backend integration are reliable.

---

## Conclusion

This PR introduces robust unit tests for both Firestore functions and the tenant form screen. It ensures reliable database operations and a smooth user experience, offering a solid testing foundation for further development.

### CODE REVIEW

**Comment by mahdiatallah** [comment]
**Overall, great progress with the recent updates!**

- Nice work removing the `@` to fix Jest issues and deleting the coverage file from the remote repo to keep things clean.

- However, a few areas could still be improved:

  1. **Add Tests for New Entries**: Could you add tests for adding a new tenant and landlord? These would help ensure full coverage and prevent potential issues.

  2. **Type Checks and Tests**: It seems there aren’t any type checks in place (e.g., ensuring the zip code is a number), nor tests for handling incorrect input types. Adding these checks and tests would make the application more robust.

  3. **Modularize Similar Tests**: Since the tests for tenants, users, and landlords have similar structures, modularizing them by extracting common logic could help reduce redundancy and improve maintainability.

These changes will further strengthen the code and test suite. Great job so far!

**Comment by baudoin-csp** [comment]
> **Overall, great progress with the recent updates!**
> 
> * Nice work removing the `@` to fix Jest issues and deleting the coverage file from the remote repo to keep things clean.
> * However, a few areas could still be improved:
>   
>   1. **Add Tests for New Entries**: Could you add tests for adding a new tenant and landlord? These would help ensure full coverage and prevent potential issues.
>   2. **Type Checks and Tests**: It seems there aren’t any type checks in place (e.g., ensuring the zip code is a number), nor tests for handling incorrect input types. Adding these checks and tests would make the application more robust.
>   3. **Modularize Similar Tests**: Since the tests for tenants, users, and landlords have similar structures, modularizing them by extracting common logic could help reduce redundancy and improve maintainability.
> 
> These changes will further strengthen the code and test suite. Great job so far!

Hello @mahdiatallah!
Thanks for your review. 

1. Regarding the add_new_tenant and landlord functions: I already implemented it in this branch but future changes in the navigation will affect my code. That's why it's not part of this PR, it will be part of the next one. On top of that, there are 500 lines to test the other functions, that's why I want to separate my code into 2 PRs.
2. You're correct about the type checks. It's one thing that we need to implement in the future. We need to use special special regex expressions across all the screens to type check email, phone number and other fields. As for now, the main goal was to obtain a functional app for M2. We'll cover edge cases for M3 as well. 
3. You're right about the redundancy in the unit tests. It's quite hard to bring modularity using the Jest framework unfortunately. I did my best to keep the code as concise as possible. If you have suggestions don't hesitate :)

**Comment by mahdiatallah** [comment]
Thanks for the detailed response!

- Understood regarding the `add_new_tenant` and `add_new_landlord` functions—splitting into two PRs makes sense, especially with the upcoming navigation changes. 
- The plan for implementing type checks and handling edge cases sounds good, and it’s great that you’re prioritizing functionality for M2 and saving these enhancements for M3.
- As for the test redundancy, I appreciate your efforts to keep things concise with Jest. If I think of any modularization suggestions, I’ll let you know.

All clear—looking forward to the next PR!

File: firebase/firebase.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@ import { setLogLevel } from "firebase/firestore";
 // https://firebase.google.com/docs/web/setup#available-libraries
 
 // Enable logs
-setLogLevel("debug");
```
[the remark this answers is not in the document]
Comment (baudoin-csp) [reply firebase/firebase.ts:12]: Yes sure!
